src/Shell/Task/ModelTask.php

```diff
@@ -395,17 +441,74 @@ public function findBelongsToMany($model, array $associations)
                 $assocTable = substr($otherTableName, 0, $otherOffset);
             }
             if ($assocTable && in_array($assocTable, $tables)) {
-                $habtmName = $this->_camelize($assocTable);
-                $assoc = [
-                    'alias' => $habtmName,
-                    'foreignKey' => $foreignKey,
-                    'targetForeignKey' => $this->_modelKey($habtmName),
-                    'joinTable' => $otherTableName
-                ];
-                if ($assoc && $this->plugin) {
-                    $assoc['className'] = $this->plugin . '.' . $assoc['alias'];
+                if (!in_array($otherTableName, $tables)) {
+                    continue;
+                }
+                $assocTableObject = TableRegistry::get($otherTableName);
+
+                $thisForeignKey = $primaryKey;
+                foreach ($thisForeignKey as $i => $fieldName) {
+                    if ($fieldName === 'id') {
+                        $thisForeignKey[$i] = Inflector::singularize($model->table()) . '_id';
+                    }
+                }
+                $otherForeignKeys = $assocTableObject->schema()->columns();
+                foreach ($otherForeignKeys as $i => $fieldName) {
+                    if ($fieldName !== 'parent_id' && (
+                        strlen($fieldName) <= 3 || !preg_match('/^.*_id$/', $fieldName))
+                    ) {
+                        unset($otherForeignKeys[$i]);
+                    }
+                }
+                $thisForeignKey = array_values($thisForeignKey);
+                $otherForeignKeys = array_values($otherForeignKeys);
+                $thisForeignKeyIntersection = array_intersect($thisForeignKey, $otherForeignKeys);
+
+                $reverseForeignKey = false;
+                $reverseForeignKeyIntersection = false;
+                $otherForeignKeysDiff = array_diff($otherForeignKeys, $thisForeignKey);
+                foreach ($otherForeignKeysDiff as $reverseForeignKey) {
```

Comment:
Hmm, I might not have understood about *composite primary keys* that you are trying to implement.

When I imagine about that, the table declaration is like the following:
```sql
CREATE TABLE lines
(
  file_id INT NOT NULL
, line_number INT NOT NULL
, PRIMARY KEY(file_id, line_number)
);
```
But this implementation seems not to support this table.

In the current implementation, the composite primary keys must contain `id` field:
```sql
CREATE TABLE lines
(
  file_id INT NOT NULL
, id INT NOT NULL -- The field for line numbers
, PRIMARY KEY(id, file_id)
);
```

So even if there is a junction table, as it contains `line_id`, your code can detect lines table:
```sql
CREATE TABLE lines_errors
(
  file_id INT NOT NULL 
, line_id INT NOT NULL -- This field allows to detect `lines` table.
, error_id INT NOT NULL
, PRIMARY KEY (file_id, line_id, error_id)
, FOREIGN KEY (file_id, line_id) REFERENCES lines(file_id, id)
, FOREIGN KEY (error_id) REFERENCES errors(id)
);
```

But it seems strange to me...